examples/cpp_classification/classification.cpp

```diff
@@ -0,0 +1,255 @@
+#include <caffe/caffe.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <iosfwd>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace caffe;  // NOLINT(build/namespaces)
+using std::string;
+
+/* Pair (label, confidence) representing a prediction. */
+typedef std::pair<string, float> Prediction;
+
+class Classifier {
+ public:
+  Classifier(const string& model_file,
+             const string& trained_file,
+             const string& mean_file,
+             const string& label_file);
+
+  std::vector<Prediction> Classify(const cv::Mat& img, int N = 5);
+
+ private:
+  void SetMean(const string& mean_file);
+
+  std::vector<float> Predict(const cv::Mat& img);
+
+  void WrapInputLayer(std::vector<cv::Mat>* input_channels);
+
+  void Preprocess(const cv::Mat& img,
+                  std::vector<cv::Mat>* input_channels);
+
+ private:
+  std::auto_ptr<Net<float> > net_;
```

Comment:
> Could you change this to using boost's smart pointer or perhaps opencv's smart pointer?

This sample is not using Boost at all, I would prefer to keep it this way especially if it's only to use a simple feature, even if Caffe already depends on Boost. Regarding OpenCV, it's uncommon to see code using their smart pointer class. Also, in this sample OpenCV should only be used for image storage and processing IMO.

>  If C++11 support is available this should be a unique_ptr. 

What do you mean? Caffe is not compiled with C++11 so this is not possible, if the project ever switches to C++11, I would be happy to use `unique_ptr` indeed.

> auto_ptr hopefully never makes it into caffe

What is wrong with auto_ptr? I can disable the copy constructor of the `Classifier` class if you are concerned about the copy semantic of auto_ptr.